lib/collection/src/common/stoppable_task.rs

```diff
@@ -28,9 +32,48 @@ impl<T> StoppableTaskHandle<T> {
         self.ask_to_stop();
         self.is_started().then_some(self.join_handle)
     }
+
+    /// Join this stoppable task and handle any panics
+    ///
+    /// Any panics are propagated through the configured panic handler. If no handler is
+    /// configured, nothing happens.
+    ///
+    /// To call this, the task must already be finished. Otherwise it panics in development, or
+    /// blocks in release.
+    pub async fn join_and_handle_panic(self) {
+        debug_assert!(
+            self.join_handle.is_finished(),
+            "Task must be finished, we cannot block here on awaiting the join handle",
+        );
+
+        match self.join_handle.await {
+            Ok(_) => {}
+            Err(err) if err.is_cancelled() => {}
+            // Propagate panic
+            Err(err) if err.is_panic() && self.panic_handler.is_some() => {
+                log::trace!("Handling stoppable task panic through custom panic handler");
+                let panic = err.into_panic();
+                let panic_handler = self.panic_handler.unwrap();
+                panic_handler(panic);
+            }
+            Err(err) if err.is_panic() => {
+                log::debug!("Stoppable task panicked without panic handler");
+            }
+            // Log error on unknown error
+            Err(err) => {
+                log::error!("Stoppable task handle error for unknown reason: {err}");
+            }
+        }
+    }
 }
 
-pub fn spawn_stoppable<F, T>(f: F) -> StoppableTaskHandle<T>
+/// Spawn stoppable task `f`
+///
+/// An optional `panic_handler` may be given, eventually called if the task panicked.
+pub fn spawn_stoppable<F, T>(
+    f: F,
+    panic_handler: Option<Box<dyn Fn(PanicPayload) + Sync + Send>>,
```

Comment:
nitpick: I think we need to adopt builder pattern more in future. There are some places which getting more and more arguments and each new argument involve a change in dozen of tests which doesn't use the param.
But in this concrete example it looks ok for now

Reply:
Agree. Good practice to pick up in the future.